### ISSUE TITLE
hwm always infinite because of zero boost?!

### ISSUE THREAD

**Issue by memeplex** [zeromq/libzmq#1596]
Despite what I set as rcv and snd hwms, I'm always getting an effective hwm == 0 because the boosts are initialized to 0 and `pipe::set_hwms` does the following:

```
void zmq::pipe_t::set_hwms (int inhwm_, int outhwm_)
{
    int in = inhwm_ + inhwmboost;
    int out = outhwm_ + outhwmboost;


    // if either send or recv side has hwm <= 0 it means infinite so we should set hwms infinite
    if (inhwm_ <= 0 || inhwmboost <= 0)
        in = 0;

    if (outhwm_ <= 0 || outhwmboost <= 0)
        out = 0;

    lwm = compute_lwm(in);
    hwm = out;
```

**Comment by memeplex** [comment]
Maybe the intention is for `boost == 0` to mean infinite boost, the same than `xxxhwm == 0` means infinite hwm, but then the initial 0 value for the boosts is wrong AFAICS.

**Comment by ricnewton** [comment]
Do you have a test case for this.  The tests seem to pass OK.  Is this an inproc socket?

The boost stuff was for an inproc socket where you have e.g. the send socket with a snd hwm of 10 and the receiving socket has a rcv hwm of 5 the out boost on the snd socket will be 5 so you get a total of 15 for the hwm.

**Comment by memeplex** [comment]
I found the problem creating xsub_nodrop tcp pubs from pyzmq with very small hwms. They would never block on send so I inspected the output of set_hwms and indeed hwm was always 0 (i.e. infinite). This is not surprising because of the `|| outhwmboost <= 0` condition, being the case that outhwmboost is 0 by default. Give me some time to concoct a coherent test from those casual observations.

**Comment by memeplex** [comment]
I've been doing so further debugging. These are more or less confirmed facts:

1. A documented behavior: setting sndhwm and/or rcvhwm *before bind/connect* has effect over the effective hwm. Setting the *after bind/connect* hasn't.
2. But... setting sndhwm and/or rcvhwm *after connect* will call `set_hwms` and hwm will probably end up zero-valued, because of the above.

The problem I was experiencing was related to setting options after bind/connect (my bad) and, after fixing that, to the size of kernel buffers that to some extent compensate for the tiny hwm of my tests.

Nevertheless I still find point 2 disturbing. First of all, another user could make the same mistake I made, setting hwm after connect; in this case, the result will be hwm = 0 instead of the safer default hwm = 1000. Second, and you know better than I about the inproc case, but is it sensible to set the default boost to infinity (supposing boost == 0 means infinity, which could be inferred from the set_hwms implementation)?

**Comment by memeplex** [comment]
Btw, the existence of `update_pipe_options` specifically for sndhwm and rcvhwm seems to suggest that, despite the docs disclaimer, these two options are allowed to be updated *after connect*. If this is the case and not only a detail of implementation for some special (inproc?) case, the point against a default infinity boost is much stronger, because without changing the default boosts any attempt to update the hwm will make the queues unbounded.

**Comment by ricnewton** [comment]
Yes we can now set sndhwm and rcvhwm after bind/connect.

When we had the old limitation of requiring the user to set hwm before connect we kept getting bug reports from confused users were setting them after connect and expecting it to work.  We were going to just return an error when they tried to do this, but it turned out just as easy to remove the limitation completely.

I believe the documentation has been updated to indicate this, maybe you are looking at old documentation? (http://api.zeromq.org/4-2:zmq-setsockopt)
